rclpy/rclpy/__init__.py

```diff
@@ -67,11 +67,15 @@ def spin_once(node):
 
     _rclpy.rclpy_wait(wait_set)
 
-    # TODO(wjwwood): properly implement this by checking the contents of the wait_set.
-    for subscription in node.subscriptions:
-        msg = _rclpy.rclpy_take(subscription.subscription_handle, subscription.msg_type)
-        if msg:
-            subscription.callback(msg)
+    sub_ready_list = _rclpy.rclpy_get_ready_subscriptions(wait_set)
+
+    for sub_ptr in sub_ready_list:
+        for sub in node.subscriptions:
+            if sub_ptr == sub.subscription_pointer:
```

Comment:
The two loops plus condition could be compressed into the following which expresses more clearly the intention:

```
for sub in [s for s in node.subscriptions if s.subscription_pointer in sub_ready_list]:
```